### PR TITLE
PyTorch Object Detector Training Implementation and Prediction Update

### PR DESCRIPTION
**Pull request by f4str**
# Description

Implementation of the `fit` method for the `PyTorchObjectDetector`, `PyTorchFasterRCNN`, and `PyTorchYolo` object detectors. This allows these models to be trained using ART. The notebook in `notebooks/poisoning_attack_bad_det_rma.ipynb` has been updated to demonstrate training these models. This is a partial implementation of #2058 as the `fit` method needs to also be implemented for the TensorFlow object detectors.

The `predict` method in the `PyTorchObjectDetector` and `PyTorchYolo` classes has been rewritten to implement the following new features:
* Cleanup and refactoring so the code is easier to read and understand.
* The `torch.no_grad()` scope is used like the `PyTorchClassifier` to prevent extraneous gradients accumulating in the model which causes slow down.
* Inputs are now batched so the entire input is not processed at once.
* The `channels_first` parameter is now used to determine whether the input needs to be transformed. The default behavior remains the same.

Additionally, the unit tests for the `test_pytorch_object_detector.py` and `test_pytorch_faster_rcnn.py` were rewritten in pytest with a new test added for model training.

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tests for `PyTorchObjectDetector` (rewritten in pytest)
- [x] Tests for `PyTorchFasterRCNN` (rewritten in pytest)
- [x] Tests for `PyTorchYolo`

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @f4str Thank you very much for your pull request. I have added a few review comments, what do you think?

File: art/estimators/object_detection/pytorch_object_detector.py [hunk 1]
```diff
@@ -356,35 +381,155 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+                prediction["scores"] = prediction_x1y1x2y2["scores"].detach().cpu().numpy()
+                if "masks" in prediction_x1y1x2y2:
+                    prediction["masks"] = prediction_x1y1x2y2["masks"].detach().cpu().numpy().squeeze()
+
+                results_list.append(prediction)
+
+        # Apply postprocessing
+        predictions = self._apply_postprocessing(preds=results_list, fit=False)
+
+        return predictions  # type: ignore
```
[the remark this answers is not in the document]
Comment (f4str) [reply art/estimators/object_detection/pytorch_object_detector.py:427]: Yes, this is necessary since since `self._apply_postprocessing` is type annotated to return `np.ndarray`. The method does not necessarily always return that type, but this is what mypy believes which causes the typing error.

File: art/estimators/object_detection/pytorch_object_detector.py [hunk 2]
```diff
@@ -356,35 +381,155 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+
+                # Form the loss function
+                loss_components = self._model(i_batch, o_batch)
+                if isinstance(loss_components, dict):
+                    loss = sum(loss_components.values())
+                else:
+                    loss = loss_components
+
+                # Do training
+                loss.backward()  # type: ignore
```
[the remark this answers is not in the document]
Comment (f4str) [reply art/estimators/object_detection/pytorch_object_detector.py:526]: Yes, this is necessary since the `sum(...)` function is type annotated to return an `int` type. This is not actually true since for a list of tensors, it returns a tensor. However, this is what mypy believes which causes the typing error.

File: art/estimators/object_detection/pytorch_yolo.py [hunk 3]
```diff
@@ -397,50 +423,161 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+                prediction["scores"] = prediction_x1y1x2y2["scores"].detach().cpu().numpy()
+                if "masks" in prediction_x1y1x2y2:
+                    prediction["masks"] = prediction_x1y1x2y2["masks"].detach().cpu().numpy().squeeze()
+
+                results_list.append(prediction)
+
+        # Apply postprocessing
+        predictions = self._apply_postprocessing(preds=results_list, fit=False)
+
+        return predictions  # type: ignore
```
[the remark this answers is not in the document]
Comment (f4str) [reply art/estimators/object_detection/pytorch_yolo.py:470]: Yes, this is necessary since since `self._apply_postprocessing` is type annotated to return `np.ndarray`. The method does not necessarily always return that type, but this is what mypy believes which causes the typing error.

File: art/estimators/object_detection/pytorch_yolo.py [hunk 4]
```diff
@@ -397,50 +423,161 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+                # Form the loss function
+                labels_t = translate_labels_art_to_yolov3(labels_art=o_batch)
+                loss_components = self._model(i_batch, labels_t)
+                if isinstance(loss_components, dict):
+                    loss = sum(loss_components.values())
+                else:
+                    loss = loss_components
+
+                # Do training
+                loss.backward()  # type: ignore
```
[the remark this answers is not in the document]
Comment (f4str) [reply art/estimators/object_detection/pytorch_yolo.py:570]: Yes, this is necessary since the `sum(...)` function is type annotated to return an `int` type. This is not actually true since for a list of tensors, it returns a tensor. However, this is what mypy believes which causes the typing error.

File: tests/estimators/object_detection/test_pytorch_faster_rcnn.py [hunk 5]
```diff
@@ -194,90 +217,68 @@ def test_errors(self):
...
-        ]
+def test_compute_loss(art_warning, get_pytorch_faster_rcnn):
+    try:
+        object_detector, x_test, y_test = get_pytorch_faster_rcnn
 
         # Compute loss
-        loss = frcnn.compute_loss(x=self.x_test, y=y)
+        loss = object_detector.compute_loss(x=x_test, y=y_test)
+
+        assert pytest.approx(0.84883332, abs=0.01) == float(loss)
```
[the remark this answers is not in the document]
Comment (f4str) [reply tests/estimators/object_detection/test_pytorch_faster_rcnn.py:276]: The input now uses the fixture which is slightly different from the original input. Therefore, it is a different output which gives a different loss value.

File: art/estimators/object_detection/pytorch_object_detector.py [hunk 6]
```diff
@@ -356,35 +377,157 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+                prediction["boxes"] = prediction_x1y1x2y2["boxes"].detach().cpu().numpy()
+                prediction["labels"] = prediction_x1y1x2y2["labels"].detach().cpu().numpy()
+                prediction["scores"] = prediction_x1y1x2y2["scores"].detach().cpu().numpy()
+                if "masks" in prediction_x1y1x2y2:
+                    prediction["masks"] = prediction_x1y1x2y2["masks"].detach().cpu().numpy().squeeze()
+
+                results_list.append(prediction)
+
+        # Apply postprocessing
+        predictions = self._apply_postprocessing(preds=results_list, fit=False)
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_object_detector.py:425]:
```suggestion
```

File: art/estimators/object_detection/pytorch_object_detector.py [hunk 7]
```diff
@@ -356,35 +377,157 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+
+        return predictions  # type: ignore
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_object_detector.py:427]:
```suggestion
        return predictions
```

File: art/estimators/object_detection/pytorch_object_detector.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/pytorch_object_detector.py:526]: Ok, thank you.

File: art/estimators/object_detection/pytorch_yolo.py [hunk 8]
```diff
@@ -397,50 +417,161 @@ def predict(self, x: np.ndarray, batch_size: int = 128, **kwargs) -> List[Dict[s
...
+                prediction["scores"] = prediction_x1y1x2y2["scores"].detach().cpu().numpy()
+                if "masks" in prediction_x1y1x2y2:
+                    prediction["masks"] = prediction_x1y1x2y2["masks"].detach().cpu().numpy().squeeze()
+
+                results_list.append(prediction)
+
+        # Apply postprocessing
+        predictions = self._apply_postprocessing(preds=results_list, fit=False)
+
+        return predictions  # type: ignore
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:470]:
```suggestion
        return predictions
```

File: art/estimators/object_detection/pytorch_yolo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/pytorch_yolo.py:570]: Ok, thank you.

File: art/estimators/object_detection/pytorch_yolo.py (hunk 8 above)
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_yolo.py:470]:
```suggestion
        return predictions
```

**Review by beat-buesser — APPROVED** [review]
@f4str Looks good to me! Thank you very much.
